### PR TITLE
New function forward_gmail_message (with tests)

### PR DESCRIPTION
**Pull request by jcwatson11**
## Summary
- Extract `_forward_gmail_message_impl` for testability
- Add 8 unit tests covering text/HTML forwarding, attachments, CC/BCC, and subject handling

## Test plan
- [x] All 8 new tests pass (`uv run pytest tests/gmail/test_forward_gmail_message.py -v`)
- [x] Existing tests unaffected

## Summary by CodeRabbit

* **New Features**
  * Enhanced the “Forward Gmail Message” tool to forward an existing email by ID, including forwarded-only usage where subject/body can be omitted.
  * Supports plain and HTML forwarding, including optional user-provided forward notes and proper formatting conversions.
  * Automatically prefixes forwarded subjects with `Fwd:` (avoids double-prefixing) and propagates To/Cc/Bcc.
  * Can include original attachments (default on) and returns confirmation with attachment details.
* **Permissions**
  * Requires Gmail read access in addition to send.
* **Tests**
  * Added comprehensive coverage for plain/HTML forwarding, formatting, subject-prefix edge cases, CC/BCC, attachment include/exclude behavior, and attachment download failure.

### CODE REVIEW

**Comment by kapilthakare-cyberpunk** [comment]
[Commenting to indicate this PR is being closed due to cleanup process]

**Comment by DannyOosterveer** [comment]
Thanks for this @jcwatson11 — a native `forward_gmail_message` fills a real gap. We hit exactly this building an email assistant on top of the MCP: reconstructing a forward from the plaintext body loses the original HTML formatting and the attachments, so a first-class forward tool is very welcome.

One observation on CI: the only failing check is `ruff-format` — the `ruff` lint check itself passes — so it looks purely cosmetic. `uv run ruff format` on the changed files should turn it green. Since "allow edits from maintainers" is enabled on this PR, a maintainer could even push that fix directly.

File: gmail/gmail_tools.py [hunk 1]
```diff
@@ -1155,6 +1155,246 @@ async def send_gmail_message(
...
+
+    forward_header_html = (
+        '<div style="color: #777;">'
+        "---------- Forwarded message ---------<br/>"
+        f"From: {original_from}<br/>"
+        f"Date: {original_date}<br/>"
+        f"Subject: {original_subject}<br/>"
+        f"To: {original_to}"
+        "</div>"
+    )
```
[the remark this answers is not in the document]
Comment (jcwatson11) [reply gmail/gmail_tools.py]: Fixed in 7441093. Header values (From/Date/Subject/To) are now run through `html.escape()` in the HTML forward block. I deliberately left `original_html` and the user-supplied `forward_message` (when `forward_message_format="html"`) un-escaped, since preserving/allowing HTML there is the intended behavior; the plain-text `forward_message` branch already escapes.

File: gmail/gmail_tools.py [hunk 2]
```diff
@@ -1155,6 +1155,246 @@ async def send_gmail_message(
...
+
+    # Construct the forward body
+    if has_html:
+        # Build HTML forward body
+        user_message_html = ""
+        if forward_message:
+            if forward_message_format == "html":
+                user_message_html = f"<div>{forward_message}</div><br/>"
```
[the remark this answers is not in the document]
Comment (jcwatson11) [reply gmail/gmail_tools.py]: Same as above — header values are now HTML-escaped as of 7441093.

File: gmail/gmail_tools.py [hunk 3]
```diff
@@ -1155,6 +1155,246 @@ async def send_gmail_message(
...
+                urlsafe_data = attachment_data.get("data", "")
+                # Convert from URL-safe to standard base64
+                standard_b64 = urlsafe_data.replace("-", "+").replace("_", "/")
+                attachments_to_send.append(
+                    {
+                        "content": standard_b64,
+                        "filename": att["filename"],
+                        "mime_type": att["mimeType"],
+                    }
+                )
```
[the remark this answers is not in the document]
Comment (jcwatson11) [reply gmail/gmail_tools.py:2293]: Good catch — fixed in 7441093. The attachment data is now padded to a multiple of 4 and decoded with `base64.urlsafe_b64decode()`, then re-encoded as standard padded base64 before it reaches `_prepare_gmail_message`'s `base64.b64decode()`. The naive `-/_` replace alone could have raised `Incorrect padding` on unpadded input.

File: gmail/gmail_tools.py [hunk 4]
```diff
@@ -1155,6 +1155,246 @@ async def send_gmail_message(
...
+                logger.info(f"[forward_gmail_message] Downloaded attachment: {att['filename']}")
+            except Exception as e:
+                logger.warning(
+                    f"[forward_gmail_message] Failed to download attachment {att['filename']}: {e}"
+                )
+
+    # Prepare the forward subject
+    forward_subject = original_subject
+    if not forward_subject.lower().startswith("fwd:"):
+        forward_subject = f"Fwd: {original_subject}"
```
[the remark this answers is not in the document]
Comment (jcwatson11) [reply gmail/gmail_tools.py]: Addressed in 7441093. The guard now strips leading whitespace and matches both `fwd:` and `fw:` so `FW: ...` no longer becomes `Fwd: FW: ...`.

File: tests/gmail/test_forward_gmail_message.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+"""
+
+import pytest
+from unittest.mock import Mock
+import base64
+import sys
+import os
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))
+
```
[the remark this answers is not in the document]
Comment (jcwatson11) [reply tests/gmail/test_forward_gmail_message.py:13]: Leaving this as-is intentionally. Every existing test suite in this repo (`tests/gappsscript`, `tests/gcontacts`, `tests/gforms`, `tests/gsheets`) uses the identical `sys.path.insert(...)` line, so matching that convention keeps this PR consistent. Switching to a `conftest.py`/editable-install approach is a reasonable improvement, but it should be a separate repo-wide change rather than a one-off here.

**Comment by jcwatson11** [comment]
Thanks @DannyOosterveer! Ran `uv run ruff format` — that was indeed the only thing failing CI, should be green now (7441093).

While I was in here I also addressed the Copilot review points in the same push: HTML-escaping the forwarded header values, normalizing attachment base64 padding so unpadded Gmail data doesn't trip `b64decode`, and broadening the subject double-prefix guard to cover `FW:`/`FWD:` variants.

File: gmail/gmail_tools.py [hunk 6]
```diff
@@ -1155,6 +1155,257 @@ async def send_gmail_message(
...
+
+    attachment_info = (
+        f" with {len(attachments_to_send)} attachment(s)" if attachments_to_send else ""
+    )
+    return f"Email forwarded{attachment_info}! Message ID: {sent_message_id}"
+
+
+@server.tool()
+@handle_http_errors("forward_gmail_message", service_type="gmail")
+@require_google_service("gmail", GMAIL_SEND_SCOPE)
```
[the remark this answers is not in the document]
Comment (jcwatson11) [reply gmail/gmail_tools.py]: Fixed in 84e33b4. Changed the decorator to `@require_google_service("gmail", ["gmail_read", GMAIL_SEND_SCOPE])`. Confirmed `_resolve_scopes` accepts a list and `"gmail_read"` maps to `GMAIL_READONLY_SCOPE` (service_decorator.py:420), matching how the other read tools in this file declare scopes. Good catch — send-only auth would have failed on the `messages().get()`/`attachments().get()` calls.

File: gmail/gmail_tools.py [hunk 7]
```diff
@@ -1155,6 +1155,257 @@ async def send_gmail_message(
...
+                        "mime_type": att["mimeType"],
+                    }
+                )
+                logger.info(
+                    f"[forward_gmail_message] Downloaded attachment: {att['filename']}"
+                )
+            except Exception as e:
+                logger.warning(
+                    f"[forward_gmail_message] Failed to download attachment {att['filename']}: {e}"
+                )
```
[the remark this answers is not in the document]
Comment (jcwatson11) [reply gmail/gmail_tools.py:2300]: Fixed in 84e33b4. Failed downloads are now collected into `failed_attachments` and, when `include_attachments=True`, the function raises with the failed filename(s) instead of silently delivering an incomplete forward. Added `test_forward_attachment_download_failure_raises` to cover it.

File: gmail/gmail_tools.py [hunk 8]
```diff
@@ -1155,6 +1155,257 @@ async def send_gmail_message(
...
+            f"<br/>"
+            f"{original_html}"
+            f"</div>"
+        )
+        body_format = "html"
+    else:
+        # Build plain text forward body
+        user_message_text = f"{forward_message}\n\n" if forward_message else ""
+        forward_body = f"{user_message_text}{forward_header_text}\n\n{original_text}"
+        body_format = "plain"
```
[the remark this answers is not in the document]
Comment (jcwatson11) [reply gmail/gmail_tools.py]: Fixed in 84e33b4. Body selection now uses `use_html = has_html or (forward_message and forward_message_format == "html")`, and when the original is plain text it's escaped and newline-converted into the HTML block. Added `test_forward_plain_original_with_html_note` to verify the HTML format is honored for plain-text originals.

**Comment by jcwatson11** [comment]
@coderabbitai Addressed all four findings in 84e33b4:

- **Read scope (Major):** decorator now requests `["gmail_read", GMAIL_SEND_SCOPE]` so the message/attachment reads are authorized, not just the send.
- **Silent partial forward (Major):** attachment download failures now abort the forward with the failed filename(s) instead of sending an incomplete message.
- **`forward_message_format="html"` for plain-text originals (Minor):** HTML body is now selected when an HTML note is supplied, with the plain-text original escaped/newline-converted into the block.
- **Test nitpick:** added `get_sent_mime_message`/`get_body_text` helpers and now decode the sent MIME to assert subject prefixing (incl. the `FW:` variant), To/Cc/Bcc, body format/content, and attachment round-trip (with an unpadded URL-safe fixture exercising the base64 normalization). New tests cover the plain-original/HTML-note and attachment-failure paths.

All 11 tests pass; `ruff format`/`ruff check` clean.

**Comment by jcwatson11** [comment]
@coderabbitai Addressed the abort-before-send nitpick in 8c073d0. `test_forward_attachment_download_failure_raises` now asserts no body-bearing `send` call is made after the raise.

Note: I assert that no send carried a `body` rather than using `assert_not_called()`, because the test harness (`create_mock_service`) invokes `.send()` once with no args while wiring the mock's return value — so `assert_not_called()` would false-fail on that setup call. Filtering `call_args_list` for the `body` kwarg isolates the real send attempt.

**Comment by taylorwilsdon** [comment]
Refactored this into the existing tools - we're at the limit for tools so we can't add new ones for specific purpose functions. @coderabbitai re-review the latest branch
